repose-aggregator/components/filters/cors/src/main/scala/org/openrepose/filters/cors/CorsFilter.scala

```diff
@@ -0,0 +1,180 @@
+    val xsdUrl: URL = getClass.getResource(SCHEMA_FILE_NAME)
+    configurationService.subscribeTo(filterConfig.getFilterName, configurationFile, xsdUrl, this, classOf[CorsConfig])
+
+    logger.trace("CORS filter initialized.")
+  }
+
+  override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain): Unit = {
+    val httpServletRequest = servletRequest.asInstanceOf[HttpServletRequest]
+    val httpServletResponse = servletResponse.asInstanceOf[HttpServletResponse]
+    val isOptions = httpServletRequest.getMethod == "OPTIONS"
+    val origin = httpServletRequest.getHeader(CommonHttpHeader.ORIGIN)
+    val requestMethodHeader = httpServletRequest.getHeader(CommonHttpHeader.ACCESS_CONTROL_REQUEST_METHOD)
+    lazy val validMethods = getValidMethodsForResource(httpServletRequest.getRequestURI)
+
+    val requestType = (Option(origin), isOptions, Option(requestMethodHeader)) match {
+        case (Some(_), true, Some(_)) => PreflightRequest
+        case (Some(_), _, None) => ActualRequest
+        case _ => NonCorsRequest
+    }
+
+    val requestedMethod = requestType match {
+      case PreflightRequest => requestMethodHeader
+      case _ => httpServletRequest.getMethod
+    }
+
+    val validationResult = requestType match {
+      case NonCorsRequest => Pass
+      case _ =>
+        (isOriginAllowed(origin), validMethods.exists(requestedMethod == _)) match {
+          case (true, true) => Pass
+          case (false, _) => OriginNotAllowed
+          case (true, false) => MethodNotAllowed
+        }
+    }
+
+    validationResult match {
+      case Pass =>
+        requestType match {
+          case NonCorsRequest => filterChain.doFilter(httpServletRequest, httpServletResponse)
+          case PreflightRequest =>
+            httpServletResponse.setHeader(CommonHttpHeader.ACCESS_CONTROL_ALLOW_CREDENTIALS, "true")
+            httpServletResponse.setHeader(CommonHttpHeader.ACCESS_CONTROL_ALLOW_ORIGIN, origin)
+            Option(httpServletRequest.getHeader(CommonHttpHeader.ACCESS_CONTROL_REQUEST_HEADERS)).foreach {
+              httpServletResponse.setHeader(CommonHttpHeader.ACCESS_CONTROL_ALLOW_HEADERS, _)}
+            getValidMethodsForResource(httpServletRequest.getRequestURI).foreach {
+              httpServletResponse.addHeader(CommonHttpHeader.ACCESS_CONTROL_ALLOW_METHODS, _)}
+            httpServletResponse.setStatus(HttpServletResponse.SC_OK)
+          case ActualRequest =>
+            filterChain.doFilter(httpServletRequest, httpServletResponse)
+            httpServletResponse.setHeader(CommonHttpHeader.ACCESS_CONTROL_ALLOW_CREDENTIALS, "true")
+            httpServletResponse.setHeader(CommonHttpHeader.ACCESS_CONTROL_ALLOW_ORIGIN, origin)
+
+            // clone the list of header names so we can add headers while we iterate through it
+            (List() ++ httpServletResponse.getHeaderNames.asScala).foreach {
```

Comment:
I'm not sure I understand the reason for cloning the header names here. Why not just iterate over the header names collection and drop the "List() ++" part?